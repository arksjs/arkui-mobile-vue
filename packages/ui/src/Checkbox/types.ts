import type { FormItemCommonProps } from '../Form/types'

export type ModelValue = number | string

export type OptionItem = {
  label: string
  value: ModelValue
}

export type UserOptionItem = ModelValue | OptionItem

export interface CheckGroupCommonProps extends FormItemCommonProps {
  options?: UserOptionItem[]
  inline?: boolean
  activeColor?: string
  value?: ModelValue | ModelValue[]
}

export interface CheckboxGroupProps extends CheckGroupCommonProps {
  value?: ModelValue[]
}

export type OnChange = (value: ModelValue[]) => void

export interface CheckboxGroupEmits {
  onChange?: OnChange
}

export interface CheckCommonProps extends FormItemCommonProps {
  value?: ModelValue
  checked?: boolean
  activeColor?: string
}

export interface CheckboxProps extends CheckCommonProps {
  circle?: boolean
}

export interface CheckCommonEmits {
  onUpdateChecked?: (checked: boolean) => void
  onCheckedChange?: (checked: boolean) => void
}

export type CheckboxEmits = CheckCommonEmits

export type {
  UserOptionItem as CheckboxOptionItem,
  UserOptionItem as RadioOptionItem
}
