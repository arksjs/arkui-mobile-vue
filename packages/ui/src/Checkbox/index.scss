@import '../style/var.scss';

.ak-checkbox {
  --ak-active-color: #{$primary-color};
  --ak-checkbox-icon-size: 20px;

  display: inline-flex;
  flex-direction: column;
  align-items: center;
  font-size: 17px;
  color: $title-color;
  text-align: left;
  position: relative;

  + .ak-checkbox {
    margin-left: 16px;
  }

  &_box {
    display: flex;
    width: 100%;
    height: 32px;
    box-sizing: border-box;
    align-items: center;
  }

  &_icon,
  &_checked-icon {
    --ak-icon-size: var(--ak-checkbox-icon-size);
    --ak-icon-color: #{$border-color};
  }

  &_checked-icon {
    display: none;

    --ak-icon-color: var(--ak-active-color);
  }

  &_input {
    position: absolute;
    left: 0;
    top: 0;
    opacity: 0;

    &:checked + .ak-checkbox_box .ak-checkbox_icon {
      display: none;
    }

    &:checked + .ak-checkbox_box .ak-checkbox_checked-icon {
      display: block;
    }
  }

  &_text {
    display: block;
    line-height: 1;
    margin-left: 12px;
  }

  &.disabled {
    opacity: 0.3;
  }
}

.ak-checkbox-group {
  display: flex;
  height: 32px;
  align-items: center;
  color: $title-color;
  box-sizing: border-box;
  flex: 1;

  &.vertical {
    flex-direction: column;
    height: auto;
    min-height: 32px;

    > .ak-checkbox {
      display: flex;
      width: 100%;

      + .ak-checkbox {
        margin-left: 0;

        &::before {
          content: '';
        }
      }
    }
  }
}
