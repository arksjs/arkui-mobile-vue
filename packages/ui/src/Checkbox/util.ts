import type { CSSProperties } from '../helpers/types'

export const getCheckStyles = (activeColor?: string) => {
  const obj: CSSProperties = {}

  activeColor && (obj['--ak-active-color'] = activeColor)
  // size != null &&
  //   size > 0 &&
  //   (obj['--ak-size'] = parseFloat(size as string) + 'px')

  return obj
}

export const getCheckClasses = (disabled?: boolean) => [
  'ak-horizontal-hairline',
  { disabled: !!disabled }
]

export const getCheckGroupClasses = ({
  inline,
  disabled
}: {
  inline?: boolean
  disabled?: boolean
}) => ({
  vertical: !inline,
  disabled: !!disabled
})
