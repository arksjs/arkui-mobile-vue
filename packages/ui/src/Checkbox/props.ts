import type { PropType } from 'vue'
import type { CheckCommonEmits, UserOptionItem } from './types'
import { formItemProps } from '../Form/form'
import { colorValidator } from '../helpers/validator'
import type { PropsToEmits } from '../helpers/types'
import { isBoolean } from '../helpers/util'

export const checkGroupProps = {
  ...formItemProps,
  options: {
    type: Array as PropType<UserOptionItem[]>,
    default: () => [] as UserOptionItem[]
  },
  inline: {
    type: Boolean,
    default: false
  },
  activeColor: {
    type: String
  }
}

export const checkEmits: PropsToEmits<CheckCommonEmits> = {
  'update:checked': checked => isBoolean(checked),
  change: checked => isBoolean(checked)
}

export const checkProps = {
  value: {
    type: [Number, String],
    default: ''
  },
  checked: {
    type: Boolean,
    default: false
  },
  disabled: {
    type: Boolean,
    default: false
  },
  activeColor: {
    type: String,
    validator: colorValidator
  },
  name: {
    type: String,
    default: ''
  }
}
