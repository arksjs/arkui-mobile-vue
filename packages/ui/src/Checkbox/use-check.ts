import { computed, onMounted, ref, watch, inject, provide } from 'vue'
import type { SetupContext } from 'vue'
import { capitalize, isStringNumberMix } from '../helpers/util'
import { useGroup, useGroupItem } from '../hooks/use-group'
import type {
  ModelValue,
  OptionItem,
  CheckGroupCommonProps,
  CheckCommonProps
} from './types'
import type { checkEmits } from './props'
import { getCheckClasses, getCheckGroupClasses, getCheckStyles } from './util'

interface GroupOptions {
  props: CheckGroupCommonProps & {
    modelValue: ModelValue | ModelValue[]
  }
  onChange: (uid: symbol) => void
}

interface GroupItem {
  uid: symbol
  getInputChecked: () => boolean
  getValue: () => number | string
  setChecked: (checked: boolean) => void
}

export function useCheck(
  props: CheckCommonProps,
  { emit }: SetupContext<typeof checkEmits>,
  name: string
) {
  const uid = Symbol()
  const groupOptions = inject<GroupOptions | null>(
    `ak${capitalize(name)}Options`,
    null
  )
  const input = ref<HTMLInputElement>()

  const name2 = computed(() => {
    return groupOptions?.props.name || props.name || ''
  })
  const disabled2 = computed(() => {
    return groupOptions?.props.disabled ?? props.disabled
  })

  function getValue() {
    return props.value ?? ''
  }

  function getInputEl() {
    return input.value as HTMLInputElement
  }

  function getInputChecked() {
    return !!getInputEl().checked
  }

  function setChecked(checked = true) {
    getInputEl().checked = checked
  }

  function onChange(e: Event) {
    if (groupOptions) {
      groupOptions.onChange(uid)
    } else {
      const checked = !!(e.target as HTMLInputElement).checked
      emit('update:checked', checked)
      emit('checkedChange', checked)
    }
  }

  watch(
    () => props.checked,
    val => {
      if (groupOptions) {
        return
      }

      const $input = getInputEl()
      const checked = !!val

      if (checked !== $input.checked) {
        $input.checked = checked
      }
    }
  )

  useGroupItem<GroupItem>(name, {
    uid,
    getInputChecked,
    getValue,
    setChecked
  })

  onMounted(() => {
    const $input = getInputEl()

    let checked: boolean
    if (groupOptions) {
      checked =
        name === 'checkbox'
          ? !!(
              Array.isArray(groupOptions.props.modelValue) &&
              props.value &&
              groupOptions.props.modelValue.includes(props.value)
            )
          : props.value === groupOptions.props.modelValue
    } else {
      checked = !!props.checked
    }

    if (checked !== $input.checked) {
      $input.checked = $input.defaultChecked = checked
    }
  })

  const classes = computed(() => getCheckClasses(disabled2.value))

  const styles = computed(() => {
    const { activeColor } = groupOptions?.props || props

    return getCheckStyles(activeColor)
  })

  return {
    input,
    name2,
    disabled2,
    onChange,
    classes,
    styles
  }
}

export function useCheckGroup<T>(
  props: CheckGroupCommonProps & {
    modelValue: T
  },
  {
    name,
    updateValue,
    watchValue
  }: {
    name: string
    updateValue: (options: {
      isChange: boolean
      children: GroupItem[]
      uid: symbol | undefined
    }) => T
    watchValue: (options: { children: GroupItem[]; value: T }) => void
  }
) {
  const root = ref<HTMLElement>()
  const { children } = useGroup<GroupItem>(name)

  function _updateValue(isChange: boolean, uid?: symbol) {
    return updateValue({ isChange, children, uid })
  }

  function onChange(uid: symbol) {
    _updateValue(true, uid)
  }

  watch(
    () => props.modelValue,
    value => {
      watchValue({ children, value })
    },
    {
      deep: true
    }
  )

  const options2 = computed(() => {
    const ret: OptionItem[] = []

    if (Array.isArray(props.options)) {
      props.options.forEach(v => {
        if (isStringNumberMix(v)) {
          ret.push({
            value: v,
            label: v.toString()
          })
        } else {
          ret.push({
            value: v.value,
            label: v.label.toString()
          })
        }
      })
    }

    return ret
  })

  onMounted(() => _updateValue(false))

  provide(`ak${capitalize(name)}Options`, {
    props,
    onChange
  })

  const classes = computed(() =>
    getCheckGroupClasses({
      inline: props.inline,
      disabled: props.disabled
    })
  )

  return {
    root,
    onChange,
    options2,
    classes
  }
}
