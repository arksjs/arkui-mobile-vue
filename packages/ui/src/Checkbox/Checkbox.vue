<template>
  <label class="ak-checkbox" :class="classes" :style="styles">
    <input
      class="ak-checkbox_input ak-form-input"
      type="checkbox"
      :name="name2"
      :value="value"
      :disabled="disabled2"
      @change="onChange"
      ref="input"
    />
    <div class="ak-checkbox_box">
      <Icon
        class="ak-checkbox_icon"
        :icon="circle ? CircleOutlined : BorderOutlined"
      />
      <Icon
        class="ak-checkbox_checked-icon"
        :icon="circle ? CheckCircleFilled : CheckSquareFilled"
      />
      <span class="ak-checkbox_text" v-if="$slots.default">
        <slot></slot>
      </span>
    </div>
  </label>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { Icon } from '../Icon'
import { checkEmits, checkProps } from './props'
import { useCheck } from './use-check'
import CircleOutlined from '../Icon/icons/CircleOutlined'
import CheckCircleFilled from '../Icon/icons/CheckCircleFilled'
import BorderOutlined from '../Icon/icons/BorderOutlined'
import CheckSquareFilled from '../Icon/icons/CheckSquareFilled'

export default defineComponent({
  name: 'ak-checkbox',
  components: { Icon },
  props: {
    ...checkProps,
    circle: {
      type: Boolean,
      default: false
    }
  },
  emits: { ...checkEmits },
  setup(props, ctx) {
    return {
      ...useCheck(props, ctx, 'checkbox'),
      CircleOutlined,
      CheckCircleFilled,
      BorderOutlined,
      CheckSquareFilled
    }
  }
})
</script>
