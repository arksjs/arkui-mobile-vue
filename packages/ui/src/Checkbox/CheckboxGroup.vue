<template>
  <div class="ta-checkbox-group" :class="classes" ref="root">
    <slot>
      <Checkbox
        v-for="item in options2"
        :key="item.value"
        :checkedValue="item.value"
        >{{ item.label }}</Checkbox
      >
    </slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, type PropType } from 'vue'
import Checkbox from './Checkbox.vue'
import {
  isStringOrNumberArray,
  cloneData,
  isSameArray,
  type PropsToEmits
} from '../helpers'
import { useCheckGroup } from './use-check'
import type { CheckboxGroupEmits, ModelValue } from './types'
import { checkGroupProps } from './props'

const isValue = (value: ModelValue[]) => isStringOrNumberArray(value)

export default defineComponent({
  name: 'ta-checkbox-group',
  components: { Checkbox },
  props: {
    ...checkGroupProps,
    modelValue: {
      type: Array as PropType<ModelValue[]>,
      validator: isValue,
      default: (): string[] => []
    }
  },
  emits: {
    'update:modelValue': isValue,
    change: isValue
  } as PropsToEmits<CheckboxGroupEmits>,
  setup(props, ctx) {
    const inputValue = ref<ModelValue[]>([])
    const { emit } = ctx

    const group = useCheckGroup(props, {
      name: 'checkbox',
      updateValue({ isChange, children }) {
        const newVal: ModelValue[] = []

        children.forEach(child => {
          if (child.getInputChecked()) {
            newVal.push(cloneData(child.getValue()))
          }
        })

        inputValue.value = newVal

        if (isChange && !isSameArray(newVal, props.modelValue)) {
          emit('update:modelValue', cloneData(newVal))
        }

        if (isChange) {
          emit('change', cloneData(newVal))
        }

        return newVal
      },
      watchValue({ children, value }) {
        if (
          isStringOrNumberArray(value) &&
          !isSameArray(value, inputValue.value)
        ) {
          const newVal: ModelValue[] = []

          children.forEach(child => {
            const checked = value.includes(child.getValue())
            child.setChecked(checked)
            checked && newVal.push(child.getValue())
          })

          inputValue.value = newVal
        }
      }
    })

    return {
      ...group
    }
  }
})
</script>
