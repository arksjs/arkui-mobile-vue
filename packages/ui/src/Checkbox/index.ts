import { multiWithInstall } from '../helpers/with-install'
import Checkbox from './Checkbox.vue'
import CheckboxGroup from './CheckboxGroup.vue'

export { Checkbox, CheckboxGroup }
export default Checkbox
export const install = multiWithInstall(Checkbox, [CheckboxGroup])
