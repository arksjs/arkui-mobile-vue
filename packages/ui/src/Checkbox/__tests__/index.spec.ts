import { describe, test, expect } from 'vitest'
import { shallowMount, mount } from '@vue/test-utils'
import { Checkbox, CheckboxGroup } from '@/Checkbox'

describe('Checkbox', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Checkbox, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('disabled=true', () => {
    const wrapper = shallowMount(Checkbox, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.classes()).toContain('disabled')
    expect(wrapper.find('input').attributes('disabled')).toBe('')
  })
})

describe('CheckboxGroup', () => {
  test('snapshot', () => {
    const wrapper = mount(CheckboxGroup, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('disabled=true', () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        disabled: true
      },
      slots: {
        default: [Checkbox]
      }
    })

    expect(wrapper.classes()).toContain('disabled')
    expect(wrapper.findComponent(Checkbox).classes()).toContain('disabled')
  })
})
