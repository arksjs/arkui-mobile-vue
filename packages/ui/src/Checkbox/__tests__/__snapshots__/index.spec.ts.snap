// Vitest Snapshot v1

exports[`Checkbox > snapshot 1`] = `
"<label class=\\"ak-checkbox ak-horizontal-hairline\\"><input class=\\"ak-checkbox_input ak-form-input\\" type=\\"checkbox\\" name=\\"\\" value=\\"\\">
  <div class=\\"ak-checkbox_box\\">
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\" class=\\"ak-checkbox_icon\\"></icon-stub>
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\" class=\\"ak-checkbox_checked-icon\\"></icon-stub>
    <!--v-if-->
  </div>
</label>"
`;

exports[`CheckboxGroup > snapshot 1`] = `"<div class=\\"ak-checkbox-group vertical\\"></div>"`;
